sync3/handler/txn_id_waiter.go

```diff
@@ -0,0 +1,96 @@
+package handler
+
+import (
+	"github.com/matrix-org/sliding-sync/sync3/caches"
+	"sync"
+	"time"
+)
+
+type TxnIDWaiter struct {
+	userID  string
+	publish func(delayed bool, update caches.Update)
+	// mu guards the queues map.
+	mu       sync.Mutex
+	queues   map[string][]*caches.RoomEventUpdate
+	maxDelay time.Duration
+}
+
+func NewTxnIDWaiter(userID string, maxDelay time.Duration, publish func(bool, caches.Update)) *TxnIDWaiter {
+	return &TxnIDWaiter{
+		userID:   userID,
+		publish:  publish,
+		mu:       sync.Mutex{},
+		queues:   make(map[string][]*caches.RoomEventUpdate),
+		maxDelay: maxDelay,
+		// TODO: metric that tracks how long events were queued for.
+	}
+}
+
+func (t *TxnIDWaiter) Ingest(up caches.Update) {
+	if t.maxDelay <= 0 {
+		t.publish(false, up)
+		return
+	}
+
+	eventUpdate, isEventUpdate := up.(*caches.RoomEventUpdate)
+	if !isEventUpdate {
+		t.publish(false, up)
+		return
+	}
+
+	ed := eventUpdate.EventData
+	// We only want to queue this event if
+	//  - our user sent it AND it lacks a txn_id; OR
+	//  - the room already has queued events.
+
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	_, roomQueued := t.queues[ed.RoomID]
+	missingTxnID := ed.Sender == t.userID && ed.TransactionID == ""
+	if !(missingTxnID || roomQueued) {
+		t.publish(false, up)
+		return
+	}
+
+	// We've decided to queue the event.
+	queue, exists := t.queues[ed.RoomID]
+	if !exists {
+		queue = make([]*caches.RoomEventUpdate, 0, 10)
+	}
+	// TODO: bound the queue size?
+	t.queues[ed.RoomID] = append(queue, eventUpdate)
+
+	// TODO: if t gets gced, will this function still run? If so, will things explode?
+	time.AfterFunc(t.maxDelay, func() { t.PublishUpToNID(ed.RoomID, ed.NID) })
```

Comment:
@kegsay could use your advice here. I guess the point is that this closure will still hold a reference to `t`, and so it `t` won't be GCed until this timer has run?